### ISSUE TITLE
Cannot load_weights with keras StringLookup layer

### ISSUE THREAD

**Issue by areiner222** [tensorflow/tensorflow#56054]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

binary

### Tensorflow Version

tf 2.8

### Custom Code

No

### OS Platform and Distribution

macOS Monterey

### Python version

3.7.13

### Current Behaviour?

```shell
I am unable to reload the vocabulary of an adapted StringLookup layer by using the {save/load}_weights api of a subclassed keras Model.
```

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
class TestModel(tf.keras.models.Model):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.sl = tf.keras.layers.StringLookup()
        
    def adapt(self, df):
        self.sl.adapt(df)
    
        
    def call(self, x):
        return self.sl(x)

inp = tf.repeat(tf.constant(['A', 'B', 'C']), 10)
df_inp = tf.data.Dataset.from_tensor_slices(inp)

test_model  = TestModel()
test_model.adapt(df_inp)
print(test_model.get_weights())

'[array([b'C', b'B', b'A'], dtype=object)]'

test_model.save_weights('tmp/check_weights')

test_model_recon = TestModel()
test_model_recon.load_weights('tmp/check_weights')
print(test_model_recon.get_weights())

'[array([], dtype=object)]'
```

### Relevant log output

_No response_</details>

**Comment by tilakrayal** [comment]
Hello @areiner222 ,
I was facing different error while trying to execute the given code.Please find the gist of it [here](https://colab.research.google.com/gist/tilakrayal/eba87b5911d50e24921856696e2749ae/untitled336.ipynb) and share all the dependencies to reproduce the issue.Thanks!

**Comment by bhack** [comment]
@tilakrayal You don't need to copy strings like `'[array([b'C', b'B', b'A'], dtype=object)]'` as these are the output:

```python
import tensorflow as tf
class TestModel(tf.keras.models.Model):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.sl = tf.keras.layers.StringLookup()
        
    def adapt(self, df):
        self.sl.adapt(df)
    
        
    def call(self, x):
        return self.sl(x)

inp = tf.repeat(tf.constant(['A', 'B', 'C']), 10)
df_inp = tf.data.Dataset.from_tensor_slices(inp)

test_model  = TestModel()
test_model.adapt(df_inp)
print(test_model.get_weights())

test_model.save_weights('tmp/check_weights')

test_model_recon = TestModel()
test_model_recon.load_weights('tmp/check_weights')
print(test_model_recon.get_weights())
```

**Comment by areiner222** [comment]
@tilakrayal Sorry about including the botched strings - I did it to quickly visualize the output without having to navigate to a gist.

If you remove the strings (like @bhack  did) and run the code, you should see the unexpected behavior.

**Comment by areiner222** [comment]
In the process of debugging, found that  replacing the  `StaticHashTable` with a `MutableHashTable` for the `lookup_table` in the  parent `IndexLookup` layer produces the expected result.

```python
import tensorflow as tf
StringLookup = tf.keras.layers.StringLookup

def _uninitialized_lookup_table(self):
  with tf.init_scope():
    return tf.lookup.experimental.MutableHashTable(
        key_dtype=self._key_dtype, value_dtype=self._value_dtype, default_value=self._default_value
      )
def _lookup_table_from_tokens(self, tokens):
  with tf.init_scope():
    lookup_table = self._uninitialized_lookup_table()
    token_start = self._token_start_index()
    token_end = token_start + tf.size(tokens)
    indices = tf.range(token_start, token_end, dtype=tf.int64)
    keys, values = (indices, tokens) if self.invert else (tokens, indices)
    lookup_table.insert(keys, values)
    return lookup_table
StringLookup._uninitialized_lookup_table = _uninitialized_lookup_table
StringLookup._lookup_table_from_tokens = _lookup_table_from_tokens

class TestModel(tf.keras.models.Model):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.sl = StringLookup()
        
    def adapt(self, df):
        self.sl.adapt(df)
    
        
    def call(self, x):
        return self.sl(x)

inp = tf.repeat(tf.constant(['A', 'B', 'C']), 10)
df_inp = tf.data.Dataset.from_tensor_slices(inp)

test_model  = TestModel()
test_model.adapt(df_inp)
print(test_model.get_weights())
test_model.save_weights('tmp/check_weights')
test_model_recon = TestModel()
test_model_recon.load_weights('tmp/check_weights')
print(test_model_recon.get_weights())
```

Output:
```
[array([b'C', b'B', b'A'], dtype=object)]
[array([b'C', b'B', b'A'], dtype=object)]
```

**Comment by bhack** [comment]
@areiner222 
As you can see we don't have serialization tests in:
https://github.com/keras-team/keras/blob/master/keras/layers/preprocessing/index_lookup_test.py

Can you try to open a PR extending that test to cover your serialization case (with something like https://github.com/keras-team/keras/blob/master/keras/layers/preprocessing/normalization_test.py#L399-L429)?

P.s. Keras PRs and Issues are submitted at:
https://github.com/keras-team/keras

**Comment by bhack** [comment]
Also `MutableHashTable` it is seems to have a `checkpoint` arg:
https://github.com/tensorflow/tensorflow/blob/0729f4133e3470bf4ea037a0d2e377409a801f16/tensorflow/python/ops/lookup_ops.py#L1832-L1834

instead `StaticHashTable` is immutable after initialization:

https://github.com/tensorflow/tensorflow/blob/3f878cff5b698b82eea85db2b60d65a2e320850e/tensorflow/python/ops/lookup_ops.py#L279-L280

See the conditions on when `MutableHashTable` or `StaticHashTable` objects are selected in the internal implementation of the parent layer:
https://github.com/keras-team/keras/blob/master/keras/layers/preprocessing/index_lookup.py

I don't know if we have some space to add an useful warning feedback to the user but I think you need to contribute a test for this.

**Comment by tilakrayal** [comment]
@areiner222 ,
Thanks for opening this issue. Development of keras moved to another [repository](https://github.com/keras-team/keras/issues).

Could you please post this issue on keras-team/keras [repo](https://github.com/keras-team/keras/issues).
To know more refer:
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999
Thank you!
